derp/derp_server.go

```diff
@@ -850,6 +989,47 @@ func (s *Server) sendServerKey(lw *lazyBufioWriter) error {
 	return err
 }
 
+func (s *Server) noteClientActivity(c *sclient) {
+	if !c.isDup.Get() {
+		// Fast path for clients that aren't in a dup set.
+		return
+	}
+	if c.isDisabled.Get() {
+		// If they're already disabled, no point checking more.
+		return
+	}
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	ds, ok := s.clients[c.key].(*dupClientSet)
+	if !ok {
+		// It became unduped in between the isDup fast path check above
+		// and the mutex check. Nothing to do.
+		return
+	}
+
+	if sh := ds.sendHistory; len(sh) != 0 && sh[len(sh)-1] == c {
+		// The client c was the last client to make activity
+		// in this set and it was already recorded. Nothing to
+		// do.
+		return
+	}
+
+	// Append this client to the list of clients who spoke last.
+	ds.sendHistory = append(ds.sendHistory, c)
+
+	// If we have a log of 3 items and see the same one 3 back
+	// being the same as the last one (pattern: A, B, A) then
+	// consider them all (probably just 2) fighting and disable
+	// them.
+	if sh := ds.sendHistory; len(sh) >= 3 && sh[len(sh)-3] == c {
+		ds.ForeachClient((*sclient).markDisabled)
+	}
+}
+
+func (c *sclient) markDisabled() { c.isDisabled.Set(true) }
```

Comment:
Used in only 1 spot directly above, fold it into that code?

Reply:
done.